### PR TITLE
Fixed #1755 - Vapor crashes with large MPAS wireframe renderer

### LINKED ISSUE #1755: Vapor crashes with large MPAS wireframe renderer
Tested on 13inch mac with 16GB ram

1. Load the data set from glade MPAS/FalkoJudt/30km
1. Add a wireframe renderer
1. Select variable `zz` (655362x75)
1. Enable the renderer

Vapor will hang and eventually crash

### PR DESCRIPTION
**Pull request by clyne**
VAPOR was crashing inside of the WireFrameRenderer with large data sets. The code (prior to this PR) was allocating more memory than necessary. This was mostly due to the brute force rendering approach taken: each edge, for each cell in the grid was drawn as a line segment. Edges shared by multiple cells (the general case) were drawn multiple times. Thus, the connectivity list could easily be 10's of GigaBytes in size. 

In addition to reducing the memory requirements by avoiding redrawing the same lines over and over, this PR takes further steps to improve performance. Namely, memory is pre-allocated for std::vector.

### CODE REVIEW

**Comment by shaomeng** [comment]
John, I guess a high-level description of why the crash occurred in the first place, and what this PR does to fix it, would be helpful for the review!

**Comment by clyne** [comment]
Good point! I added a description at the top of the PR.

> On Oct 2, 2019, at 2:29 PM, Samuel Li <notifications@github.com> wrote:
> 
> John, I guess a high-level description of why the crash occurred in the first place, and what this PR does to fix it, would be helpful for the review!
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

John Clyne
National Center for Atmospheric Research
303.497.1236 (w), 303.809.1922 (c)
clyne@ucar.edu

File: include/vapor/WireFrameRenderer.h [hunk 1]
```diff
@@ -81,17 +81,74 @@ class RENDER_API WireFrameRenderer : public Renderer
...
+			_maxEntries(maxEntries),
+			_maxLinesPerVertex(maxLinesPerVertex)
+		 {}	
+
+		bool InList(size_t idx0, size_t idx1) {
+			VAssert(idx0 < _maxEntries);
+			VAssert(idx1 < _maxEntries);
+
+			if (idx1 < idx0) {
+				size_t tmp = idx0;
```
[the remark this answers is not in the document]
Comment (clyne) [reply include/vapor/WireFrameRenderer.h]: Done.

File: lib/render/WireFrameRenderer.cpp [hunk 2]
```diff
@@ -113,47 +114,238 @@ bool WireFrameRenderer::_isCacheDirty() const
...
+
+		size_t idx1 = nodeMap[cellNodeIndices[(i+1)%count]];
+		VAssert(idx1 != invalidIndex);
+
+		// Don't draw line segment if it's already been drawn
+		//
+		if (drawList.InList(idx0,idx1)) continue;
+
+        indices.push_back(idx0);
+        indices.push_back(idx1);
```
[the remark this answers is not in the document]
Comment (clyne) [reply lib/render/WireFrameRenderer.cpp]: Done

**Comment by clyne** [comment]
The color mapping regression has been fixed, and the mismatch between Grid and OpenGL buffer precision is now handled more robustly.

File: include/vapor/WireFrameRenderer.h [hunk 3]
```diff
@@ -81,17 +82,72 @@ class RENDER_API WireFrameRenderer : public Renderer
...
+					_drawList[idx0*_maxLinesPerVertex+i] = idx1;
+					return(false);
+				}
+			}
+			return(false);
+		}
+	private:
+		vector<GLuint> _drawList;
+		size_t _maxEntries;
+		size_t _maxLinesPerVertex;
```
Comment (shaomeng) [inline include/vapor/WireFrameRenderer.h]: If these member variables aren't gonna change, then give them a const qualifier. 
```suggestion
        const size_t _maxEntries;
		const size_t _maxLinesPerVertex;
```

**Pull request merged**